clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -1304,6 +1310,19 @@ private CompletedFetch initializeCompletedFetch(CompletedFetch nextCompletedFetc
         return completedFetch;
     }
 
+    private void handleOffsetOutOfRange(long fetchOffset,
+                                        TopicPartition topicPartition,
+                                        String reason) {
+        if (subscriptions.hasDefaultOffsetResetPolicy()) {
+            log.info("Fetch offset {} is out of range for partition {}, resetting offset",
+                topicPartition, fetchOffset);
+            subscriptions.requestOffsetReset(topicPartition);
+        } else {
+            throw new OffsetOutOfRangeException(Collections.singletonMap(
```

Comment:
Another improvement we can make here is to add the fetch offset to the exception message. Also, I'm considering if we should log this event even if we throw the exception back to the user. Otherwise, the user application might swallow it and we won't know it happened.